File: frontend/src/types/developer.ts
export interface DeveloperRunSummary {
  id: string
  status: string
  workflow: string
  provider: string
  model: string
  step_count: number
  tool_call_count: number
  model_call_count: number
  total_tokens: number
  estimated_cost_microusd: number
  duration_ms: number
  error_code: string | null
  started_at: string
}

export interface AgentStep {
  sequence: number
  node: string
  status: string
  input_summary: Record<string, unknown>
  output_summary: Record<string, unknown>
  duration_ms: number
  error_code: string | null
}

export interface DeveloperRunDetail {
  run: DeveloperRunSummary & {
    user_query: string
    answer: string | null
    evidence_refs: string[]
    tool_names: string[]
    agent_steps: AgentStep[]
    model_calls: Array<{
      sequence: number
      status: string
      provider: string
      model: string
      latency_ms: number
      total_tokens: number
      error_code: string | null
    }>
  }
  tool_traces: Array<{
    evidence_id: string
    tool_name: string
    status: string
    duration_ms: number
    error_code: string | null
    started_at: string
  }>
  plans: Array<{
    version: number
    status: string
    tasks: Array<{ task_id: string; agent: string; tool_name: string; depends_on: string[] }>
    validation_error_codes: string[]
    replan_reason: string | null
  }>
  memory_accesses: Array<{
    status: string
    matched_memory_ids: string[]
    reason: string
    created_at: string
  }>
}

export interface EvaluationSummary {
  id: string
  comparison_group_id: string | null
  dataset_version: string
  workflow: string
  provider: string
  model: string
  status: string
  case_count: number
  passed_count: number
  tool_selection_accuracy: number
  numeric_accuracy: number
  task_completion_rate: number
  evidence_coverage: number
  hallucination_rate: number
  routing_accuracy: number
  average_handoff_count: number
  average_latency_ms: number
  p50_latency_ms: number
  p95_latency_ms: number
  total_tokens: number
  estimated_cost_microusd: number
  started_at: string
}

export interface EvaluationComparison {
  comparison_group_id: string
  dataset_version: string
  items: EvaluationSummary[]
}

export interface FailureSample {
  id: string
  case_id: string
  workflow: string
  error_stage: string
  error_code: string | null
  created_at: string
}

File: backend/app/schemas/developer.py
"""Developer-only trace and evaluation response schemas."""

from datetime import datetime

from pydantic import BaseModel, Field

from app.schemas.agent import AgentRunResponse


class DeveloperRunSummary(BaseModel):
    id: str
    status: str
    workflow: str
    provider: str
    model: str
    step_count: int
    tool_call_count: int
    model_call_count: int
    total_tokens: int
    estimated_cost_microusd: int
    duration_ms: int
    error_code: str | None
    started_at: datetime


class DeveloperRunListResponse(BaseModel):
    items: list[DeveloperRunSummary]
    total: int = Field(ge=0)
    page: int = Field(ge=1)
    page_size: int = Field(ge=1, le=100)


class DeveloperToolTrace(BaseModel):
    evidence_id: str
    tool_name: str
    status: str
    duration_ms: int
    error_code: str | None
    started_at: datetime


class DeveloperPlanTask(BaseModel):
    task_id: str
    agent: str
    tool_name: str
    depends_on: list[str]


class DeveloperPlanTrace(BaseModel):
    version: int
    status: str
    tasks: list[DeveloperPlanTask]
    validation_error_codes: list[str]
    replan_reason: str | None
    created_at: datetime


class DeveloperMemoryAccess(BaseModel):
    status: str
    matched_memory_ids: list[str]
    reason: str
    created_at: datetime


class DeveloperRunDetail(BaseModel):
    run: AgentRunResponse
    tool_traces: list[DeveloperToolTrace]
    plans: list[DeveloperPlanTrace]
    memory_accesses: list[DeveloperMemoryAccess]


class DeveloperEvaluationSummary(BaseModel):
    id: str
    comparison_group_id: str | None
    dataset_version: str
    workflow: str
    provider: str
    model: str
    status: str
    case_count: int
    passed_count: int
    tool_selection_accuracy: float
    numeric_accuracy: float
    task_completion_rate: float
    evidence_coverage: float
    hallucination_rate: float
    routing_accuracy: float
    average_handoff_count: float
    average_latency_ms: int
    p50_latency_ms: int
    p95_latency_ms: int
    total_tokens: int
    estimated_cost_microusd: int
    started_at: datetime


class DeveloperEvaluationListResponse(BaseModel):
    items: list[DeveloperEvaluationSummary]
    total: int = Field(ge=0)


class DeveloperComparison(BaseModel):
    comparison_group_id: str
    dataset_version: str
    items: list[DeveloperEvaluationSummary]


class DeveloperComparisonListResponse(BaseModel):
    items: list[DeveloperComparison]
    total: int = Field(ge=0)


class DeveloperFailureSample(BaseModel):
    id: str
    evaluation_run_id: str
    dataset_version: str
    case_id: str
    workflow: str
    error_stage: str
    error_code: str | None
    summary: dict
    created_at: datetime


class DeveloperFailureListResponse(BaseModel):
    items: list[DeveloperFailureSample]
    total: int = Field(ge=0)
